packages-exp/messaging-exp/src/helpers/register.ts

```diff
@@ -22,17 +22,27 @@ import {
   InstanceFactory
 } from '@firebase/component';
 import { ERROR_FACTORY, ErrorCode } from '../util/errors';
+import { isSwSupported, isWindowSupported } from '../api/isSupported';
 
 import { MessagingService } from '../messaging-service';
 import { _registerComponent } from '@firebase/app-exp';
-import { isSupported } from './isSupported';
 
-const messagingFactory: InstanceFactory<'messaging-exp'> = (
+const WindowMessagingFactory: InstanceFactory<'messaging-exp'> = (
   container: ComponentContainer
 ) => {
-  if (!isSupported()) {
-    throw ERROR_FACTORY.create(ErrorCode.UNSUPPORTED_BROWSER);
-  }
+  // Conscious decision to make this async check non-blocking during the messaging instance
+  // initialization phase for performance consideration. An error would be thrown latter for
+  // developer's information. Developers can then choose to import and call `isSupported` for
+  // special handling.
+  isWindowSupported()
+    .then(isSupported => {
+      if (!isSupported) {
+        throw ERROR_FACTORY.create(ErrorCode.UNSUPPORTED_BROWSER);
+      }
+    })
+    .catch(_ => {
+      throw ERROR_FACTORY.create(ErrorCode.INDEXED_DB_UNSUPPORTED);
+    });
```

Comment:
This seems this code will never hit because `isWindowSupported()` & `isSwSupported ()` never throws.

Reply:
good catch!